Wrap plugin list output content

didrocks opened the pull request:
Wrap plugin list output content up to 120 characters depending on terminal size.

Added tests to ensure if detection fails, we are using the 120 characters. as well. Fixes https://bugs.launchpad.net/snapcraft/+bug/1587057.

Note that it's using put cols as a subprocess. A couple of years ago, I had to look into how to get the terminal size on different platforms in a reliable way, and it seems it's the best one without duplicating a fragile logic which evolves over time as new terminals arise.

didrocks commented:
Note: I saw the bug report is more about column alignement than wrapping text. This can be used as a base. I'll look later this week if there is an easy way to reuse existing components for this (so, please do not merge yet)

sergiusens commented on snapcraft/main.py:
```diff
@@ -117,9 +120,22 @@ def _get_version():
+        with suppress(subprocess.CalledProcessError):
+            # this is the only way to get current terminal size reliably
+            # without duplicating a bunch of logic
+            command = ['tput', 'cols']
```
Surprisingly I would of thought python had something built in for this

didrocks replied:
Isn't it? :)
I remember to have been surprised by this as well a year ago, when looking for it, and after some digging into the web, that was the only way to get this done. I just did a quick look again yesterday, and nothing changed in between

sergiusens replied:
Seems we need to add `ncurses-bin` to `debian/control`

sergiusens commented:
Thanks for getting this started! As requested I will wait.

didrocks commented:
I just rebased on latest master and added a couple of commits to now have the "by column" functionality (there is no built-in module to do that), but the logic was quite easy to implement with the required tests of course :)

sergiusens commented:
retest this please

sergiusens commented on snapcraft/tests/test_commands_list_plugins.py:
```diff
@@ -23,25 +24,45 @@
+        def raise_error(x):
+            raise OSError()
+        mock_subprocess.side_effect = raise_error
+        main(['list-plugins'])
```
oh, there's a similar test in `integration_tests` reason for adt to fail most likely

`runtests.sh integration` ftw ;-)

Or `python3 -m unittest integration_tests ...`

sergiusens commented:
Nice tests!

didrocks commented:
Rebase on master to have correct coverage report (which should increase) and rerunning autopkgtests

elopio commented on snapcraft/tests/test_commands_list_plugins.py:
```diff
@@ -23,25 +24,45 @@
+    # plugin list when wrapper at MAX_CHARACTERS_WRAP
+    default_plugin_output = (
+        "ant        catkin  copy  jdk     kernel  maven  nodejs   python3  tar-content\n"  # noqa
+        "autotools  cmake   go    kbuild  make    nil    python2  scons  \n")  # noqa
```
For consistency, please use single quotes for strings.
And it hurts me seeing python code with more than 80 lines, I'm weird that way. Would it be too ugly to split the string before the \n?
'ant        catkin  copy  jdk     kernel  maven  nodejs   '
'python3  tar-content\n'
'autotools  cmake   go    kbuild  make    nil    '
'python2  scons  \n'

didrocks replied:
I'm a fan of 120 characters, but let's not open this debate here :p

Here, I used the #noqa stenza as the important part is to be able to check visually as well that we have a column output, making clear the test is about that. I have no strong opinion if you think it's still readable when being broken in multiple lines, just tell me :)

elopio commented on snapcraft/tests/test_commands_list_plugins.py:
```diff
@@ -1,6 +1,6 @@
 # -*- Mode:Python; indent-tabs-mode:nil; tab-width:4 -*-
 #
-# Copyright (C) 2015 Canonical Ltd
+# Copyright (C) 2015-2016 Canonical Ltd
```
legal nit: this should be 2015, 2016

elopio commented on snapcraft/tests/test_common.py:
```diff
@@ -101,3 +101,45 @@ def test_arch_triplet_migration_message(self):
+
+    elements_list = ["ant", "autotools", "catkin", "cmake", "copy", "go",
+                     "jdk", "kbuild", "kernel", "make", "maven", "nil",
+                     "nodejs", "python2", "python3", "scons", "tar-content"]
```
single quotes please.

elopio commented on snapcraft/tests/test_common.py:
```diff
@@ -101,3 +101,45 @@ def test_arch_triplet_migration_message(self):
+                     "nodejs", "python2", "python3", "scons", "tar-content"]
+
+    def test_format_output_in_columns_default(self):
+        '''Format output on 2 lines, with default max-width and space sep'''
```
double quotes for comments :p

didrocks commented:
Just pushed the fixes (minus the 80 characters one, waiting on your feedback), thanks for the review!

didrocks commented:
The last commit should fix the autopkgtest failure (it does at least when running adt-run locally) :)

didrocks commented:
it seems that adt running in jenkins has other variables set that me running adt-run (with qemu) directly, do you know the difference?

The output is:
testtools.matchers._impl.MismatchError: 'ant\nautotools\ncatkin\ncmake\ncopy\ngo\njdk\nkbuild\nkernel\nmake\nmaven\nnil\nnodejs\npython2\npython3\nscons\ntar-content' not in 'ant        catkin  copy  jdk     kernel  maven  nodejs   python3  tar-content\nautotools  cmake   go    kbuild  make    nil    python2  scons  \n'

Which kinds of puzzle me as this is the old output (and it can only happen if tput cols returns less than 20). It seems to me that it's the older snapcraft version which runs, are we sure that the adt runner is building and running a new snapcraft package first (as we are using SNAPCRAFT_FROM_INSTALLED=1 using the "snapcraft" command)?

didrocks commented:
I fixed the second integration test, but it seems to me to be mostly a duplicate of the first one (running in a project directory though, so unsure if that influences, but there is no local plugin or plugin difference). Maybe time to drop it?

sergiusens commented:
El 01/06/16 a las 03:42, Didier Roche escribió:
> Just pushed the fixes (minus the 80 characters one, waiting on your
> feedback), thanks for the review!

elopio won this battle long ago :-P
I just keep QA happy fwiw ;-)

didrocks commented:
pushed commit to keep QA happy :)

elopio commented:
I saw your IRC message and I lol'd.

I'm happy :+1:

elopio commented:
Also +1 on dropping that test_main.py

sergiusens commented:
If the integration_test is useless, let's drop it!

If `ncurses-bin` is indeed necessary, please add, feel free to merge then or ping me and I'll merge.

sergiusens commented:
oh, this is an essential package, nvm
